In [1]:
import pandas as pd


In [2]:
fake_data= pd.read_csv("data/Fake.csv")
fake_data.head()


title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   
4   Pope Francis Just Called Out Donald Trump Dur...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   
2  On Friday, it was revealed that former Milwauk...    News   
3  On Christmas day, Donald Trump announced that ...    News   
4  Pope Francis used his annual Christmas Day mes...    News   

                date  
0  December 31, 2017  
1  December 31, 2017  
2  December 30, 2017  
3  December 29, 2017  
4  December 25, 2017

In [3]:
True_data=pd.read_csv("data/True.csv")
True_data.head()

title  \
0  As U.S. budget fight looms, Republicans flip t...   
1  U.S. military to accept transgender recruits o...   
2  Senior U.S. Republican senator: 'Let Mr. Muell...   
3  FBI Russia probe helped by Australian diplomat...   
4  Trump wants Postal Service to charge 'much mor...   

                                                text       subject  \
0  WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1  WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2  WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3  WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4  SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   

                 date  
0  December 31, 2017   
1  December 29, 2017   
2  December 31, 2017   
3  December 30, 2017   
4  December 29, 2017


lable the data fake =0 and true =1

In [4]:
fake_data['lable']=0
True_data['lable']=1

In [5]:
fake_data.shape ,True_data.shape

((23481, 5), (21417, 5))

### Do you need to balance it?  ((23481, 5), (21417, 5))

Not strictly necessary — we can train directly.

Many ML models (Naive Bayes, Logistic Regression) can handle slight imbalance.

If we leave it as-is, just check performance per class (precision, recall, F1-score) so our  model isn’t biased toward the larger class.

In [6]:
fake_data.head()


title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   
4   Pope Francis Just Called Out Donald Trump Dur...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   
2  On Friday, it was revealed that former Milwauk...    News   
3  On Christmas day, Donald Trump announced that ...    News   
4  Pope Francis used his annual Christmas Day mes...    News   

                date  lable  
0  December 31, 2017      0  
1  December 31, 2017      0  
2  December 30, 2017      0  
3  December 29, 2017      0  
4  December 25, 2017      0

In [7]:
True_data.head()

title  \
0  As U.S. budget fight looms, Republicans flip t...   
1  U.S. military to accept transgender recruits o...   
2  Senior U.S. Republican senator: 'Let Mr. Muell...   
3  FBI Russia probe helped by Australian diplomat...   
4  Trump wants Postal Service to charge 'much mor...   

                                                text       subject  \
0  WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1  WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2  WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3  WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4  SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   

                 date  lable  
0  December 31, 2017       1  
1  December 29, 2017       1  
2  December 31, 2017       1  
3  December 30, 2017       1  
4  December 29, 2017       1

combine the fake and true data 

In [8]:
data=pd.concat([fake_data,True_data],ignore_index=True)
data=data.sample(frac=1 ,random_state=42).reset_index(drop=True)
data.head()

title  \
0  Ben Stein Calls Out 9th Circuit Court: Committ...   
1  Trump drops Steve Bannon from National Securit...   
2  Puerto Rico expects U.S. to lift Jones Act shi...   
3   OOPS: Trump Just Accidentally Confirmed He Le...   
4  Donald Trump heads for Scotland to reopen a go...   

                                                text       subject  \
0  21st Century Wire says Ben Stein, reputable pr...       US_News   
1  WASHINGTON (Reuters) - U.S. President Donald T...  politicsNews   
2  (Reuters) - Puerto Rico Governor Ricardo Rosse...  politicsNews   
3  On Monday, Donald Trump once again embarrassed...          News   
4  GLASGOW, Scotland (Reuters) - Most U.S. presid...  politicsNews   

                  date  lable  
0    February 13, 2017      0  
1       April 5, 2017       1  
2  September 27, 2017       1  
3         May 22, 2017      0  
4       June 24, 2016       1

This code combines fake and real news datasets, randomly shuffles all samples to avoid order bias, resets the index, and displays the first five rows for verification.

In [9]:
data.isnull().sum()

title      0
text       0
subject    0
date       0
lable      0
dtype: int64

 #### Has no miss value so not need work handel miss value  data set is cleaned data set not need cleaning 


### Text preprocessing
Convert all text to lowercase

Remove punctuation, numbers, or unnecessary symbols

In [11]:
text_cols = ['title', 'text']  


for col in text_cols:
    data[col] = data[col].str.lower()


data.head()


title  \
0  ben stein calls out 9th circuit court: committ...   
1  trump drops steve bannon from national securit...   
2  puerto rico expects u.s. to lift jones act shi...   
3   oops: trump just accidentally confirmed he le...   
4  donald trump heads for scotland to reopen a go...   

                                                text       subject  \
0  21st century wire says ben stein, reputable pr...       US_News   
1  washington (reuters) - u.s. president donald t...  politicsNews   
2  (reuters) - puerto rico governor ricardo rosse...  politicsNews   
3  on monday, donald trump once again embarrassed...          News   
4  glasgow, scotland (reuters) - most u.s. presid...  politicsNews   

                  date  lable  
0    February 13, 2017      0  
1       April 5, 2017       1  
2  September 27, 2017       1  
3         May 22, 2017      0  
4       June 24, 2016       1

#### Remove unwanted elements

In [12]:
import re

def clean_text(text):
    # Remove punctuation and numbers
    text = re.sub(r'[^a-zA-Z\s]', '', text)
    # Remove extra spaces
    text = re.sub(r'\s+', ' ', text).strip()
    return text

# Apply to both title and text columns
data['title'] = data['title'].apply(clean_text)
data['text'] = data['text'].apply(clean_text)


### Remove stopwords

Stopwords are common words like “the, is, and” which usually don’t help the model because they do not give meaning  Why we do this Removing stopwords improves:
##### ✔ text cleaning
##### ✔ machine learning model accuracy
##### ✔ reduces noise
##### ✔ reduces feature size 

In [13]:
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS

def remove_stopwords(text):
    return ' '.join([word for word in text.split() if word not in ENGLISH_STOP_WORDS])

data['title'] = data['title'].apply(remove_stopwords)
data['text'] = data['text'].apply(remove_stopwords)


#### Stemming or lemmatization 
it will conver words to root form

In [14]:
from nltk.stem import PorterStemmer
ps =PorterStemmer()
def stem_text(text):
    return ' '.join([ps.stem(word) for word in text.split()])
data['title'] = data['title'].apply(stem_text)
data['text'] = data['text'].apply(stem_text)
data.head() 

title  \
0  ben stein call th circuit court commit coup dt...   
1       trump drop steve bannon nation secur council   
2     puerto rico expect lift jone act ship restrict   
3  oop trump just accident confirm leak isra inte...   
4      donald trump head scotland reopen golf resort   

                                                text       subject  \
0  st centuri wire say ben stein reput professor ...       US_News   
1  washington reuter presid donald trump remov ch...  politicsNews   
2  reuter puerto rico governor ricardo rossello s...  politicsNews   
3  monday donald trump embarrass countri accident...          News   
4  glasgow scotland reuter presidenti candid abro...  politicsNews   

                  date  lable  
0    February 13, 2017      0  
1       April 5, 2017       1  
2  September 27, 2017       1  
3         May 22, 2017      0  
4       June 24, 2016       1

### Combine title and text
Most Fake News models work better if you combine the title and text into one column:
WHY combine?  Because:

Title is short but very powerful (high meaning)

Text has more details
 When we combine, the model learns both

If separate → the model misses some meaning.

If combined → more context, better accuracy.

In [15]:
data['content'] = data['title'] + ' ' + data['text']


#### Remove title and text

We combined the title and text into one column called **content** so the model can use all the information together. After combining, we removed the original title and text columns to avoid duplication and save memory. Using only one column also makes the model faster and easier to train. If we keep the old columns, nothing breaks, but the data becomes bigger and may reduce model accuracy.


In [16]:
data = data.drop(['title', 'text'], axis=1)
data.head()

subject                 date  lable  \
0       US_News    February 13, 2017      0   
1  politicsNews       April 5, 2017       1   
2  politicsNews  September 27, 2017       1   
3          News         May 22, 2017      0   
4  politicsNews       June 24, 2016       1   

                                             content  
0  ben stein call th circuit court commit coup dt...  
1  trump drop steve bannon nation secur council w...  
2  puerto rico expect lift jone act ship restrict...  
3  oop trump just accident confirm leak isra inte...  
4  donald trump head scotland reopen golf resort ...

In [20]:
if 'date' in data.columns:
    data.drop(columns=['date'], inplace=True)


The date column was removed because it is not useful for classification and only adds unnecessary data.Therefore, the date column was dropped to simplify the data and reduce memory usage.

In [21]:
data.to_csv("cleaned_data.csv", index=False)


In [22]:
import pickle

with open("cleaned_data.pkl", "wb") as f:
    pickle.dump(data, f)


In [23]:
print("Saved rows:", data.shape)
data.head()


Saved rows: (44898, 3)


subject  lable                                            content
0       US_News      0  ben stein call th circuit court commit coup dt...
1  politicsNews      1  trump drop steve bannon nation secur council w...
2  politicsNews      1  puerto rico expect lift jone act ship restrict...
3          News      0  oop trump just accident confirm leak isra inte...
4  politicsNews      1  donald trump head scotland reopen golf resort ...